src/Processors/Executors/ExecutingGraph.cpp

```diff
@@ -390,17 +402,25 @@ bool ExecutingGraph::updateNode(uint64_t pid, Queue & queue, Queue & async_queue
     return true;
 }
 
-void ExecutingGraph::cancel()
+void ExecutingGraph::cancel(bool cancel_all_processors)
 {
     std::exception_ptr exception_ptr;
 
     {
         std::lock_guard guard(processors_mutex);
-        for (auto & processor : *processors)
+        uint64_t num_processors = processors->size();
+        for (uint64_t proc = 0; proc < num_processors; ++proc)
         {
             try
             {
-                processor->cancel();
+                /// Stop all processors in the general case, but in a specific case
+                /// where the pipeline needs to return a result on a partially read table,
+                /// stop only the processors that read from the source
+                if (cancel_all_processors || source_processors.at(proc))
+                {
+                    IProcessor * processor = processors->at(proc).get();
+                    processor->cancel();
+                }
```

Comment:
Also, a call to `finish()` on the output ports of the source processors for the case of partial cancel is needed here to avoid troubles in some corner cases. For example, we actually have no guarantees for an arbitrary processor to cancel after `cancel()` is called, so calling `port.finish()` could improve the situation in such a scenario.

Reply:
Currently for stopping queries with `read_overflow_mode = 'break'` setting there is only `cancel()` call without `finish()` for ports https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/Executors/ExecutionThreadContext.cpp#L49.  Should we also call `port.finish()` in this case?

Reply:
Yes, let's try to add `port.finish()`. I would expect that without it reading in most cases (e.g. from `system.numbers`) will not stop, cause cancel is not implemented for it.

Finishing in case of `break` mode is not needed. We stop pipeline anyway. `cancel()` for processor just means that we want to stop currently executing job (and only implemented for some processors) to stop pipeline earlier.

Reply:
All right, I forgot about this code
https://github.com/ClickHouse/ClickHouse/blob/6509d9e525c89987f14ae290962c865072499406/src/Processors/ISource.cpp#L45
so, actually cancel() is implemented for all the sources. In this case, we don't need to add `port.finish()`. (Well, we can I suppose, but it looks not significant)

Reply:
Just want to clarify, are there any source processors that are not inherited from `ISource`? If there are no such processes then ports will call `finish()` automatically. At the end of `work()` function there is a check for a `cancel` flag https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/ISource.cpp#L107 which marks processor as finished. And on the next `prepare()` it will call `finish()` for output port https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/ISource.cpp#L24.

Reply:
Oh, I see previous comment. Haven't reload a page for too long :)